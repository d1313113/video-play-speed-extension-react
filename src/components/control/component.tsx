import React, { FC } from "react";
import "./styles.scss";
import { Slider, Button } from "antd";

const RATES = [1, 2, 3, 4, 5];

interface ControlProps {
  videoList: number[];
  onSetRate: (idx: number, value: number) => Promise<void>;
}

export const Control: FC<ControlProps> = (props: ControlProps) => {
  const handleSetVideoRate = (index: number, value: number): void => {
    props.onSetRate(index, value);
  };

  const handleClickRate = (index: number, rate: number): void => {
    props.onSetRate(index, rate);
  };

  const { videoList } = props;

  return (
    <main className="control">
      <div className="title">可拖动step调整速度</div>
      <section className="control-main">
        {videoList.map((vRate, index) => (
          <div className="video-item" key={index}>
            <Slider
              onChange={(value: number): void =>
                handleSetVideoRate(index, value)
              }
              min={0.1}
              max={5}
              step={0.1}
              defaultValue={1}
              value={vRate}
            />
            <div className="rate-wrapper">
              {RATES.map(rate => (
                <Button
                  key={rate}
                  onClick={(): void => handleClickRate(index, rate)}
                  size="small"
                  type={rate === 1 ? "primary" : "default"}
                  disabled={rate === vRate}
                >
                  {rate === 1 ? "Reset" : `✖️${rate}倍`}
                </Button>
              ))}
            </div>
          </div>
        ))}
      </section>
    </main>
  );
};
