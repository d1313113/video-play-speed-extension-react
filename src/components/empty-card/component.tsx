import React, { FC } from "react";
import { Empty } from "antd";
import "./styles.scss";

export const EmptyCard: FC = () => (
  <section className="empty-wrapper">
    <Empty className="empty" description="当前页面未找到视频:)" />
  </section>
);
