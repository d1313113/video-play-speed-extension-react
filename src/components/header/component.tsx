import React, { FC } from "react";
import "./styles.scss";

export const Header: FC = () => {
  return (
    <header className="header-wrapper">
      <div className="title">视频播放速度调节插件</div>
      <div className="sub-title">可以让你轻松调整视频播放速度</div>
    </header>
  );
};
