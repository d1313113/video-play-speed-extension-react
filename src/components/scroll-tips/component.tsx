import React, { FC } from "react";
import { DoubleRightOutlined } from "@ant-design/icons";
import "./styles.scss";

export const ScrollTips: FC = () => (
  <section className="tips-wrapper">
    <div className="tips">Scroll to load more</div>
    <DoubleRightOutlined
      className="icon-tips"
      rotate={90}
      style={{ fontSize: "20px" }}
    />
  </section>
);
