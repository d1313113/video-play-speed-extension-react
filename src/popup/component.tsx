import React, { FC, useEffect, useState } from "react";
import { browser } from "webextension-polyfill-ts";
import { throttle } from "lodash";
import getCurrentTab from "@/utils/getCurrentTab";
import {
  GET_VIDEOS_PLAYBACK_RATE,
  SET_VIDEO_PLAYBACK_RATE,
  $$
} from "@/content-script/actions/videosPlaybackRate";

import { Header } from "@/components/header";
import { EmptyCard } from "@/components/empty-card";
import { Control } from "@/components/control";
import { ScrollTips } from "@/components/scroll-tips";
import "./styles.scss";

export const Popup: FC = () => {
  const [videoList, setVideoList] = useState<number[]>([]);
  const [scrollTop, setScrollTop] = useState<number>(0);

  const initPopUp = async (): Promise<void> => {
    const { id: tabId } = await getCurrentTab();
    if (tabId != null) {
      const resp /** number[] */ = await browser.tabs.sendMessage(tabId, {
        action: GET_VIDEOS_PLAYBACK_RATE
      });
      setVideoList(resp);
    }
  };

  const handleSetVideoRate = async (
    idx: number,
    value: number
  ): Promise<void> => {
    const { id: tabId } = await getCurrentTab();

    if (tabId != null) {
      await browser.tabs.sendMessage(tabId, {
        action: SET_VIDEO_PLAYBACK_RATE,
        idx,
        value
      });
      initPopUp();
    }
  };

  const handleScroll = (html: HTMLElement): void => {
    try {
      const { scrollTop } = html;
      setScrollTop(scrollTop);
    } catch (e) {
      console.error(e);
    }
  };

  // Sends the `popupMounted` event
  useEffect(() => {
    browser.runtime.sendMessage({ popupMounted: true });
    initPopUp();
    const html = $$<HTMLElement>("html")[0];
    const _throttle = throttle((): void => handleScroll(html), 100);
    window.addEventListener("scroll", _throttle, false);

    return (): void => {
      window.removeEventListener("scroll", _throttle, false);
    };
  }, []);

  // Renders the component tree
  return (
    <div className="popup-container">
      <div className="container">
        <Header />
        {!videoList.length ? (
          <EmptyCard />
        ) : (
          <Control videoList={videoList} onSetRate={handleSetVideoRate} />
        )}
        {!!videoList.length && videoList.length > 5 && scrollTop < 30 && (
          <ScrollTips />
        )}
      </div>
    </div>
  );
};
