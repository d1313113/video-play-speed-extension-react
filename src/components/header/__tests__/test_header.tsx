import * as React from "react";
import { Header } from "../component";
import renderer, { ReactTestRendererJSON } from "react-test-renderer";

it("component renders", () => {
  const tree:
    | ReactTestRendererJSON
    | ReactTestRendererJSON[]
    | null = renderer.create(<Header />).toJSON();
  expect(tree).toMatchSnapshot();
});
