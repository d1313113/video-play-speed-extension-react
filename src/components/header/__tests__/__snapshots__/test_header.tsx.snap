// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`component renders 1`] = `
<header
  className="header-wrapper"
>
  <div
    className="title"
  >
    视频播放速度调节插件
  </div>
  <div
    className="sub-title"
  >
    可以让你轻松调整视频播放速度
  </div>
</header>
`;
